File: android/app/src/main/java/com/algonit/algo/features/chat/presentation/components/ConfirmationCard.kt
package com.algonit.algo.features.chat.presentation.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.algonit.algo.features.chat.data.model.Message

/**
 * Action confirmation card displayed when the assistant proposes
 * a write operation (e.g., pausing a campaign, sending a message).
 *
 * Shows a warning icon, the action description, and Confirm/Cancel buttons.
 * Destructive actions are highlighted with error colors.
 *
 * @param message The message containing the confirmation details.
 * @param onConfirm Callback when the user confirms the action.
 * @param onCancel Callback when the user cancels the action.
 */
@Composable
fun ConfirmationCard(
    message: Message,
    onConfirm: () -> Unit,
    onCancel: () -> Unit,
    modifier: Modifier = Modifier
) {
    val actionName = message.structuredData?.action ?: "Action"
    val isDestructive = actionName.contains("DELETE", ignoreCase = true) ||
            actionName.contains("REMOVE", ignoreCase = true) ||
            actionName.contains("CANCEL", ignoreCase = true)

    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = if (isDestructive) {
                MaterialTheme.colorScheme.errorContainer.copy(alpha = 0.3f)
            } else {
                MaterialTheme.colorScheme.tertiaryContainer.copy(alpha = 0.3f)
            }
        ),
        shape = MaterialTheme.shapes.medium
    ) {
        Column(
            modifier = Modifier.padding(16.dp)
        ) {
            // Header with warning icon
            Row(
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    imageVector = Icons.Default.Warning,
                    contentDescription = "Action requires confirmation",
                    tint = if (isDestructive) {
                        MaterialTheme.colorScheme.error
                    } else {
                        MaterialTheme.colorScheme.tertiary
                    },
                    modifier = Modifier.size(20.dp)
                )

                Spacer(modifier = Modifier.width(8.dp))

                Text(
                    text = "Confirmation Required",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold,
                    color = if (isDestructive) {
                        MaterialTheme.colorScheme.error
                    } else {
                        MaterialTheme.colorScheme.tertiary
                    }
                )
            }

            Spacer(modifier = Modifier.height(8.dp))

            // Action description
            Text(
                text = formatActionName(actionName),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface
            )

            // Target details from structured data
            message.structuredData?.target?.let { target ->
                Spacer(modifier = Modifier.height(4.dp))
                target.entries.forEach { (key, value) ->
                    val displayKey = key.replace("_", " ")
                        .replaceFirstChar { it.uppercase() }
                    Text(
                        text = "$displayKey: $value",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }

            Spacer(modifier = Modifier.height(16.dp))

            // Action buttons
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp, Alignment.End)
            ) {
                OutlinedButton(
                    onClick = onCancel,
                    colors = ButtonDefaults.outlinedButtonColors(
                        contentColor = MaterialTheme.colorScheme.error
                    )
                ) {
                    Text("Cancel")
                }

                Button(
                    onClick = onConfirm,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = if (isDestructive) {
                            MaterialTheme.colorScheme.error
                        } else {
                            MaterialTheme.colorScheme.primary
                        }
                    )
                ) {
                    Text("Confirm")
                }
            }
        }
    }
}

/**
 * Converts an action identifier like "PAUSE_CAMPAIGN" into
 * a readable string like "Pause Campaign".
 */
private fun formatActionName(action: String): String {
    return action
        .replace("_", " ")
        .lowercase()
        .split(" ")
        .joinToString(" ") { word ->
            word.replaceFirstChar { it.uppercase() }
        }
}

File: ios/Algo/Features/Chat/Views/ConfirmationCard.swift
// ConfirmationCard.swift
// Algo
//
// Action confirmation card displayed within assistant messages when the
// AI proposes a write operation (pause campaign, send email, etc.).
// Shows action details and Confirm/Cancel buttons.

import SwiftUI

// MARK: - ConfirmationCard

struct ConfirmationCard: View {

    // MARK: - Properties

    /// Optional structured data containing action details and target info.
    let structuredData: StructuredData?

    /// Callback invoked when the user confirms the action.
    let onConfirm: () -> Void

    /// Callback invoked when the user cancels the action.
    let onCancel: () -> Void

    /// Optional countdown timer for expiring confirmations.
    var expiresIn: TimeInterval?

    @State private var remainingTime: Int = 0
    @State private var isExpired = false

    // MARK: - Body

    var body: some View {
        VStack(alignment: .leading, spacing: AlgoTheme.Spacing.sm) {
            // Header with warning icon
            headerRow

            // Action details
            if let details = actionDetails {
                detailSection(details)
            }

            // Target info (campaign name, status, etc.)
            if let target = structuredData?.target {
                targetSection(target)
            }

            // Expiration countdown
            if let _ = expiresIn, !isExpired {
                expirationIndicator
            }

            // Action buttons
            if !isExpired {
                buttonRow
            } else {
                expiredNotice
            }
        }
        .padding(AlgoTheme.Spacing.md)
        .background(AlgoTheme.Colors.warning.opacity(0.06))
        .clipShape(RoundedRectangle(cornerRadius: AlgoTheme.CornerRadius.md))
        .overlay(
            RoundedRectangle(cornerRadius: AlgoTheme.CornerRadius.md)
                .strokeBorder(AlgoTheme.Colors.warning.opacity(0.3), lineWidth: 1)
        )
        .onAppear {
            startExpirationTimer()
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Action confirmation required")
    }

    // MARK: - Header

    private var headerRow: some View {
        HStack(spacing: AlgoTheme.Spacing.xs) {
            Image(systemName: isDestructive ? "exclamationmark.triangle.fill" : "checkmark.shield.fill")
                .font(.system(size: AlgoTheme.IconSize.md))
                .foregroundStyle(isDestructive ? AlgoTheme.Colors.warning : AlgoTheme.Colors.primary)

            Text("Action Confirmation")
                .font(AlgoTypography.labelLarge)
                .foregroundStyle(AlgoTheme.Colors.textPrimary)

            Spacer()
        }
    }

    // MARK: - Details

    private func detailSection(_ details: String) -> some View {
        Text(details)
            .font(AlgoTypography.bodySmall)
            .foregroundStyle(AlgoTheme.Colors.textSecondary)
    }

    // MARK: - Target

    private func targetSection(_ target: [String: JSONValue]) -> some View {
        VStack(alignment: .leading, spacing: AlgoTheme.Spacing.xxs) {
            ForEach(sortedTargetPairs(target), id: \.key) { pair in
                HStack {
                    Text(pair.key.replacingOccurrences(of: "_", with: " ").capitalized)
                        .font(AlgoTypography.caption)
                        .foregroundStyle(AlgoTheme.Colors.textTertiary)

                    Spacer()

                    Text(pair.value)
                        .font(AlgoTypography.captionBold)
                        .foregroundStyle(AlgoTheme.Colors.textPrimary)
                }
            }
        }
        .padding(AlgoTheme.Spacing.xs)
        .background(AlgoTheme.Colors.secondaryBackground)
        .clipShape(RoundedRectangle(cornerRadius: AlgoTheme.CornerRadius.sm))
    }

    // MARK: - Expiration

    private var expirationIndicator: some View {
        HStack(spacing: AlgoTheme.Spacing.xxs) {
            Image(systemName: "clock")
                .font(.system(size: 12))

            Text("Expires in \(remainingTime)s")
                .font(AlgoTypography.caption)
        }
        .foregroundStyle(remainingTime <= 10 ? AlgoTheme.Colors.error : AlgoTheme.Colors.textTertiary)
    }

    private var expiredNotice: some View {
        Text("This confirmation has expired.")
            .font(AlgoTypography.bodySmall)
            .foregroundStyle(AlgoTheme.Colors.textTertiary)
            .italic()
    }

    // MARK: - Buttons

    private var buttonRow: some View {
        HStack(spacing: AlgoTheme.Spacing.sm) {
            // Cancel button
            Button(action: onCancel) {
                HStack(spacing: AlgoTheme.Spacing.xxs) {
                    Image(systemName: "xmark")
                        .font(.system(size: 12, weight: .bold))
                    Text("Cancel")
                }
                .font(AlgoTypography.labelMedium)
                .foregroundStyle(AlgoTheme.Colors.error)
                .frame(maxWidth: .infinity)
                .padding(.vertical, AlgoTheme.Spacing.xs)
                .background(AlgoTheme.Colors.error.opacity(0.1))
                .clipShape(RoundedRectangle(cornerRadius: AlgoTheme.CornerRadius.sm))
            }
            .accessibilityLabel("Cancel action")

            // Confirm button
            Button(action: onConfirm) {
                HStack(spacing: AlgoTheme.Spacing.xxs) {
                    Image(systemName: "checkmark")
                        .font(.system(size: 12, weight: .bold))
                    Text("Confirm")
                }
                .font(AlgoTypography.labelMedium)
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .padding(.vertical, AlgoTheme.Spacing.xs)
                .background(AlgoTheme.Colors.success)
                .clipShape(RoundedRectangle(cornerRadius: AlgoTheme.CornerRadius.sm))
            }
            .accessibilityLabel("Confirm action")
        }
    }

    // MARK: - Helpers

    /// Extracts a human-readable description of the pending action.
    private var actionDetails: String? {
        guard let action = structuredData?.action else { return nil }
        return action.replacingOccurrences(of: "_", with: " ").capitalized
    }

    /// Whether this action is potentially destructive (delete, pause, etc.).
    private var isDestructive: Bool {
        guard let action = structuredData?.action?.lowercased() else { return false }
        return action.contains("delete") || action.contains("pause") || action.contains("remove")
    }

    /// Converts target dictionary into sorted display pairs.
    private func sortedTargetPairs(_ target: [String: JSONValue]) -> [(key: String, value: String)] {
        target.compactMap { key, value -> (String, String)? in
            switch value {
            case .string(let s): return (key, s)
            case .int(let i): return (key, "\(i)")
            case .double(let d): return (key, String(format: "%.2f", d))
            case .bool(let b): return (key, b ? "Yes" : "No")
            default: return nil
            }
        }
        .sorted { $0.key < $1.key }
    }

    /// Starts the expiration countdown timer.
    private func startExpirationTimer() {
        guard let expiresIn, expiresIn > 0 else { return }

        remainingTime = Int(expiresIn)

        Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { timer in
            Task { @MainActor in
                remainingTime -= 1
                if remainingTime <= 0 {
                    timer.invalidate()
                    isExpired = true
                }
            }
        }
    }
}

// MARK: - Preview

#Preview("Confirmation Cards") {
    ScrollView {
        VStack(spacing: 20) {
            ConfirmationCard(
                structuredData: StructuredData(
                    type: StructuredData.DataType.actionConfirmation,
                    metrics: nil,
                    chartType: nil,
                    timeRange: nil,
                    action: "PAUSE_CAMPAIGN",
                    status: nil,
                    target: [
                        "campaign_name": .string("Spring Sale"),
                        "status": .string("active"),
                        "today_impressions": .int(12345)
                    ],
                    campaignId: nil
                ),
                onConfirm: { print("Confirmed") },
                onCancel: { print("Cancelled") }
            )

            ConfirmationCard(
                structuredData: StructuredData(
                    type: StructuredData.DataType.actionConfirmation,
                    metrics: nil,
                    chartType: nil,
                    timeRange: nil,
                    action: "DELETE_CAMPAIGN",
                    status: nil,
                    target: ["campaign_name": .string("Old Campaign")],
                    campaignId: nil
                ),
                onConfirm: {},
                onCancel: {},
                expiresIn: 30
            )
        }
        .padding()
    }
}
